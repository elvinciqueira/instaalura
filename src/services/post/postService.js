import {parseCookies} from 'nookies'
import HttpClient from '../../infra/HttpClient'
import {LOGIN_COOKIE_APP_TOKEN} from '../login/loginService'

import {isStagingEnv} from '../../infra/env/isStagingEnv'

const BASE_URL = isStagingEnv
  ? // Back End de DEV
    'https://instalura-api.vercel.app'
  : // Back End de PROD
    'https://instalura-api.vercel.app'

export const postService = () => {
  const cookies = parseCookies()
  const token = cookies[LOGIN_COOKIE_APP_TOKEN]

  return {
    async createPost(
      {photoUrl, description, filter},
      HttpClientModule = HttpClient,
    ) {
      try {
        const response = await HttpClientModule(`${BASE_URL}/api/posts`, {
          method: 'POST',
          headers: {
            Authorization: `Bearer ${token}`,
          },
          body: {
            photoUrl,
            description,
            filter,
          },
        })

        return response.data
      } catch (err) {
        throw new Error('Não foi possível criar um novo post')
      }
    },

    async likeDislike(postId, HttpClientModule = HttpClient) {
      const response = await HttpClientModule(
        `${BASE_URL}/api/posts/${postId}/like`,
        {
          method: 'POST',
          headers: {
            Authorization: `Bearer ${token}`,
          },
          body: {},
        },
      )

      return response.data
    },

    async getPosts(path, HttpClientModule = HttpClient) {
      try {
        const url = `${BASE_URL}/${path}`

        const response = await HttpClientModule(url, {
          headers: {
            authorization: `Bearer ${token}`,
          },
        })

        return response.data
      } catch (e) {
        throw new Error('Não foi possível pegar os posts')
      }
    },
  }
}
