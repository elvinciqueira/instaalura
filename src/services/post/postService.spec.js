import {postService} from './postService'

const data = {
  photoUrl: 'image-url',
  description: 'awesome description',
  filter: 'instagram-filter',
}

async function HttpClientModule() {
  return {
    data: [data],
  }
}

const HttpClientModuleError = jest.fn()

describe('postService', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('createPost()', () => {
    describe('when user create a new post', () => {
      describe('and succeed', () => {
        test('return post data', async () => {
          const responseData = [data]
          const response = await postService().createPost(
            {
              photoUrl: 'image-url',
              description: 'awesome description',
              filter: 'instagram-filter',
            },
            HttpClientModule,
          )

          expect(response).toEqual(responseData)
        })
      })
    })

    describe('and fails', () => {
      test('throws an error', async () => {
        HttpClientModuleError.mockImplementationOnce(() =>
          Promise.reject('Não foi possível criar o post'),
        )

        try {
          await postService().createPost(
            {
              photoUrl: 'image-url',
              description: 'awesome description',
              filter: 'instagram-filter',
            },
            HttpClientModuleError,
          )
        } catch (error) {
          expect(error).toMatchSnapshot()
        }

        expect(HttpClientModuleError).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('getPosts()', () => {
    describe('when user try to get new posts', () => {
      describe('and succeed', () => {
        test('return all posts', async () => {
          const responseData = [data]
          const response = await postService().getPosts(
            'api/users/post',
            HttpClientModule,
          )

          expect(response).toEqual(responseData)
        })
      })

      describe('and fails', () => {
        test('throws an error', async () => {
          HttpClientModuleError.mockImplementationOnce(() =>
            Promise.reject('Não foi possível carregar os posts'),
          )

          try {
            await postService().getPosts(
              'api/users/post',
              HttpClientModuleError,
            )
          } catch (error) {
            expect(error).toMatchSnapshot()
          }

          expect(HttpClientModuleError).toHaveBeenCalledTimes(1)
        })
      })
    })
  })
})
