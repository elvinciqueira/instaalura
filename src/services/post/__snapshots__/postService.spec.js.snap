// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postService createPost() and fails throws an error 1`] = `[Error: Não foi possível criar um novo post]`;

exports[`postService getPosts() when user try to get new posts and fails throws an error 1`] = `[Error: Não foi possível pegar os posts]`;
